### ISSUE TITLE
Segfault after loading swap file with Recover.vim

### ISSUE THREAD

**Issue by dequis** [neovim/neovim#1836]
Recover.vim is https://github.com/chrisbra/Recover.vim

This looks like it might be very specific to my setup, not sure, i'm just going to dump the details of how it happens right now and try to narrow it down later.

```
[dx@bolsa test/bitlbee/bitlbee]$ vim protocols/account.c
```

This results in no apparent response, it looks like it's dead. Pressing enter does nothing. Pressing ctrl-c gets the usual swap recovery screen:

![](http://dump.dequis.org/2sv0h.png)

Pressing enter or ctrl-c makes it segfault:

![](http://dump.dequis.org/yzZGf.png)

If i press two ctrl-c fast enough, the contents of the file are shown for roughly half a second and then it segfaults.

![](http://dump.dequis.org/yg1R9.png)

```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007fc0781c39ae in realloc () from /usr/lib/libc.so.6
(gdb)
(gdb) bt
#0  0x00007fc0781c39ae in realloc () from /usr/lib/libc.so.6
#1  0x00000000004e6175 in xrealloc (ptr=0x7fff00000000, size=<optimized out>) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/memory.c:129
#2  0x000000000056e3ab in dynamic_buffer_ensure (buf=buf@entry=0x7fff9b772f50, desired=<optimized out>) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/os/shell.c:276
#3  0x000000000056e579 in system_data_cb (rstream=0x16ef480, data=<optimized out>, eof=<optimized out>) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/os/shell.c:286
#4  0x000000000056ba8f in read_cb (rstream=<optimized out>, data=0x16f3f70, eof=<optimized out>) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/os/job.c:442
#5  0x000000000056d79d in read_cb (stream=0x16f42d0, cnt=3, buf=<optimized out>) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/os/rstream.c:364
#6  0x00007fc078c24b94 in ?? () from /usr/lib/libuv.so.1
#7  0x00007fc078c254dc in ?? () from /usr/lib/libuv.so.1
#8  0x00007fc078c29cda in uv.io_poll () from /usr/lib/libuv.so.1
#9  0x00007fc078c1d3f7 in uv_run () from /usr/lib/libuv.so.1
#10 0x000000000056c9e0 in loop (run_mode=UV_RUN_ONCE) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/os/event.c:196
#11 0x000000000056ccc4 in event_poll (ms=ms@entry=3569) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/os/event.c:122
#12 0x000000000056db11 in input_poll (ms=ms@entry=4000) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/os/input.c:301
#13 0x000000000056db87 in inbuf_poll (ms=4000) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/os/input.c:317
#14 0x000000000056debb in os_inchar (buf=buf@entry=0x7f9d41 <typebuf_init+1> "", maxlen=maxlen@entry=87, ms=ms@entry=-1, tb_change_cnt=tb_change_cnt@entry=2) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/os/input.c:100
#15 0x00000000004a6888 in inchar (buf=0x7f9d41 <typebuf_init+1> "", maxlen=263, wait_time=-1, tb_change_cnt=2) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/getchar.c:2497
#16 0x00000000004a8680 in vgetorpeek (advance=advance@entry=1) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/getchar.c:2317
#17 0x00000000004a888b in vgetc () at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/getchar.c:1396
#18 0x00000000004a8c6f in safe_vgetc () at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/getchar.c:1516
#19 0x000000000051a6f5 in normal_cmd (oap=oap@entry=0x7fff9b7754d0, toplevel=toplevel@entry=true) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/normal.c:491
#20 0x000000000050e49c in main_loop (cmdwin=cmdwin@entry=0, noexmode=noexmode@entry=0) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/main.c:748
#21 0x000000000051052d in main (argc=<optimized out>, argv=<optimized out>) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/main.c:534
```

The [valgrind output](http://dump.dequis.org/wUU3W.txt) seems uninteresting but running it under valgrind stops it from crashing. Still gets the black screen but doesn't crash afterwards.

Backtrace when attaching gdb to neovim at the point where i need to press ctrl-c:

```
(gdb) bt
#0  0x00007f2cc86540d9 in syscall () from /usr/lib/libc.so.6
#1  0x00007f2cc90545ea in uv.epoll_wait () from /usr/lib/libuv.so.1
#2  0x00007f2cc9052e4a in uv.io_poll () from /usr/lib/libuv.so.1
#3  0x00007f2cc90463f7 in uv_run () from /usr/lib/libuv.so.1
#4  0x000000000056c9e0 in loop (run_mode=UV_RUN_ONCE) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/os/event.c:196
#5  0x000000000056ccc4 in event_poll (ms=ms@entry=-1) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/os/event.c:122
#6  0x000000000056c1f3 in job_wait (job=job@entry=0x2371ca0, ms=ms@entry=-1) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/os/job.c:305
#7  0x000000000056e9dc in shell (cmd=0x236d520 "TERM=vt100 LC_ALL=C vim -u NONE -es -V  'protocols/account.c'", extra_args=extra_args@entry=0x0, input=input@entry=0x0, len=0, output=output@entry=0x7fff27c58348,
    nread=nread@entry=0x7fff27c58350, silent=true, forward_output=false) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/os/shell.c:242
#8  0x000000000056ebd5 in os_system (cmd=<optimized out>, input=input@entry=0x0, len=<optimized out>, output=output@entry=0x7fff27c58348, nread=nread@entry=0x7fff27c58350)
    at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/os/shell.c:176
#9  0x0000000000454131 in get_system_output_as_rettv (argvars=0x7fff27c58470, rettv=0x7fff27c58ab0, retlist=retlist@entry=false) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/eval.c:14641
#10 0x0000000000454203 in f_system (argvars=<optimized out>, rettv=<optimized out>) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/eval.c:14689
#11 0x0000000000459faf in call_func (funcname=funcname@entry=0x237695b "system(cmd)", len=len@entry=6, rettv=rettv@entry=0x7fff27c58ab0, argcount=argcount@entry=1, argvars=argvars@entry=0x7fff27c58470, firstline=firstline@entry=1,
    lastline=1, doesrange=0x7fff27c58654, evaluate=1, selfdict=0x0) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/eval.c:6945
#12 0x000000000045d2a5 in get_func_tv (name=0x237695b "system(cmd)", len=6, rettv=rettv@entry=0x7fff27c58ab0, arg=arg@entry=0x7fff27c58a58, firstline=1, lastline=1, doesrange=0x7fff27c58654, evaluate=1, selfdict=0x0)
    at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/eval.c:6779
#13 0x000000000045b9b1 in eval7 (arg=arg@entry=0x7fff27c58a58, rettv=rettv@entry=0x7fff27c58ab0, evaluate=evaluate@entry=1, want_string=want_string@entry=0) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/eval.c:4123
#14 0x000000000045bbb7 in eval6 (arg=arg@entry=0x7fff27c58a58, rettv=rettv@entry=0x7fff27c58ab0, evaluate=evaluate@entry=1, want_string=want_string@entry=0) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/eval.c:3833
#15 0x000000000045be3d in eval5 (arg=arg@entry=0x7fff27c58a58, rettv=rettv@entry=0x7fff27c58ab0, evaluate=evaluate@entry=1) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/eval.c:3685
#16 0x000000000045c13c in eval4 (arg=arg@entry=0x7fff27c58a58, rettv=rettv@entry=0x7fff27c58ab0, evaluate=evaluate@entry=1) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/eval.c:3429
#17 0x000000000045c9d2 in eval3 (arg=arg@entry=0x7fff27c58a58, rettv=rettv@entry=0x7fff27c58ab0, evaluate=evaluate@entry=1) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/eval.c:3349
#18 0x000000000045caed in eval2 (arg=arg@entry=0x7fff27c58a58, rettv=rettv@entry=0x7fff27c58ab0, evaluate=evaluate@entry=1) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/eval.c:3286
#19 0x000000000045cc05 in eval1 (arg=arg@entry=0x7fff27c58a58, rettv=rettv@entry=0x7fff27c58ab0, evaluate=evaluate@entry=1) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/eval.c:3219
#20 0x000000000045d7ef in eval0 (arg=0x237695b "system(cmd)", rettv=rettv@entry=0x7fff27c58ab0, nextcmd=nextcmd@entry=0x7fff27c58b48, evaluate=1) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/eval.c:3181
#21 0x00000000004622e6 in ex_let (eap=0x7fff27c58b40) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/eval.c:1370
#22 0x00000000004f00a2 in do_one_cmd (cmdlinep=cmdlinep@entry=0x7fff27c591c8, sourcing=sourcing@entry=1, cstack=cstack@entry=0x7fff27c58d20, fgetline=fgetline@entry=0x462e22 <get_func_line>, cookie=cookie@entry=0x2371470)
    at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/ex_docmd.c:1919
#23 0x00000000004f0a16 in do_cmdline (cmdline=<optimized out>, cmdline@entry=0x0, fgetline=fgetline@entry=0x462e22 <get_func_line>, cookie=cookie@entry=0x2371470, flags=flags@entry=7)
    at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/ex_docmd.c:621
#24 0x00000000004598b9 in call_user_func (fp=fp@entry=0x2374230, argcount=argcount@entry=0, argvars=argvars@entry=0x7fff27c595c0, rettv=rettv@entry=0x7fff27c597a0, firstline=firstline@entry=1, lastline=lastline@entry=1, selfdict=0x0)
    at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/eval.c:18731
#25 0x0000000000459efa in call_func (funcname=funcname@entry=0x2376c60 "recover#ConfirmSwapDiff", len=len@entry=23, rettv=rettv@entry=0x7fff27c597a0, argcount=argcount@entry=0, argvars=argvars@entry=0x7fff27c595c0,
    firstline=firstline@entry=1, lastline=1, doesrange=0x7fff27c5979c, evaluate=1, selfdict=0x0) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/eval.c:6920
#26 0x000000000045d2a5 in get_func_tv (name=name@entry=0x2376c60 "recover#ConfirmSwapDiff", len=23, rettv=rettv@entry=0x7fff27c597a0, arg=arg@entry=0x7fff27c597b8, firstline=1, lastline=1, doesrange=0x7fff27c5979c, evaluate=1,
    selfdict=0x0) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/eval.c:6779
#27 0x0000000000460b3d in ex_call (eap=0x7fff27c59850) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/eval.c:2694
#28 0x00000000004f00a2 in do_one_cmd (cmdlinep=cmdlinep@entry=0x7fff27c59ed8, sourcing=sourcing@entry=1, cstack=cstack@entry=0x7fff27c59a30, fgetline=fgetline@entry=0x48cf57 <getnextac>, cookie=cookie@entry=0x7fff27c59f80)
    at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/ex_docmd.c:1919
#29 0x00000000004f0a16 in do_cmdline (cmdline=<optimized out>, cmdline@entry=0x0, fgetline=fgetline@entry=0x48cf57 <getnextac>, cookie=cookie@entry=0x7fff27c59f80, flags=flags@entry=7)
    at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/ex_docmd.c:621
#30 0x000000000048c520 in apply_autocmds_group (event=event@entry=EVENT_SWAPEXISTS, fname=0x2333f20 "/home/dx/test/bitlbee/bitlbee/protocols/account.c", fname_io=fname_io@entry=0x0, force=force@entry=0, group=group@entry=-3,
    buf=buf@entry=0x0, eap=0x0) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/fileio.c:6740
#31 0x000000000048c8a9 in apply_autocmds (event=event@entry=EVENT_SWAPEXISTS, fname=<optimized out>, fname_io=fname_io@entry=0x0, force=force@entry=0, buf=buf@entry=0x0)
    at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/fileio.c:6356
#32 0x0000000000484c3f in do_swapexists (buf=buf@entry=0x205cf30, fname=fname@entry=0x2333f80 "/home/dx/.vim/tmp/account.c.swp") at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/memline.c:3232
#33 0x00000000004893a8 in findswapname (buf=buf@entry=0x205cf30, dirp=dirp@entry=0x7fff27c5a538, old_fname=old_fname@entry=0x0) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/memline.c:3382
#34 0x000000000048975c in ml_open_file (buf=0x205cf30) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/memline.c:512
#35 0x0000000000489875 in check_need_swap (newfile=newfile@entry=1) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/memline.c:556
#36 0x000000000048e662 in readfile (fname=0x205f290 "protocols/account.c", sfname=<optimized out>, from=from@entry=0, lines_to_skip=lines_to_skip@entry=0, lines_to_read=lines_to_read@entry=2147483647, eap=eap@entry=0x0, flags=1)
    at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/fileio.c:621
#37 0x00000000004b28e7 in open_buffer (read_stdin=read_stdin@entry=0, eap=eap@entry=0x0, flags=flags@entry=0) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/buffer.c:153
#38 0x000000000050e946 in create_windows (parmp=parmp@entry=0x7fff27c5a8f0) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/main.c:1701
#39 0x0000000000510316 in main (argc=<optimized out>, argv=<optimized out>) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/main.c:446
```
Frame `#25` suggests blaming [Recover.vim](https://github.com/chrisbra/Recover.vim)

Distro: Arch linux
[.nvimrc](http://dump.dequis.org/uyZOf.txt)
[neovim-git PKGBUILD modified for debug symbols](http://dump.dequis.org/UuHhK.txt)
[/home/dx/test/bitlbee/bitlbee/protocols/account.c](http://dump.dequis.org/IwHfo.c)
[/home/dx/.vim/tmp/account.c.swp](http://dump.dequis.org/PwUxO.swp)

**Comment by frangio** [comment]
I believe this isn't a problem with Recover.vim. With a fresh build and an empty .nvimrc I also seem unable to recover from swap files. I get an `E309: Unable to read block 1 from .test.swp` error. With my usual configuration, instead of showing this error neovim simply segfaults. I don't know why this happens.

**Comment by justinmk** [comment]
@frangio Was the swap file created by Vim, by chance? Does it happen on swap files create by Neovim?

**Comment by dequis** [comment]
I haven't had issues with swap files since I removed Recover.vim

Also I just recovered a swap file dated "Mon Dec 29 08:34:57 2014", that's a few weeks before i started using neovim (my ~/.vim/tmp directory is full of crap).

I also haven't bothered updating the git version that I installed originally, bff48b2, so if @frangio's bug is a recent regression, i won't be able to reproduce it.

**Comment by frangio** [comment]
It happens only with Neovim. I've been doing some more tests and I found it's a more subtle bug than I thought. It only happens with `set directory^=~/.nvim/swp//`, and *only* when trying to recover through Neovim's "found existing swap file" dialog. In fact, `nvim -r ~/.nvim/swp/[…]test.swp` recovers correctly. Recover through Vim works fine, both letting it find the existing swap file and with the `-r` option.

What's weird is that without changing the `directory` option, that is having swap files in the local directory, everything works fine.

To reproduce I open an existing file, modify it, and execute `:preserve`. Then kill the `nvim` process via `kill -9`, and afterwards try to recover.

**Comment by justinmk** [comment]
Nice repro, I will try to get a dump.

**Comment by frangio** [comment]
I think the `E309: Unable to read block 1 from .test.swp` error I was getting earlier was due to killing Neovim too soon and not executing `:preserve` first. Just to be clear.

**Comment by justinmk** [comment]
@frangio What OS are you on? I wonder if you have an old version of libuv, which had a bug until we upgraded in the last couple months. Did you try:

    rm -rf build/ .deps/ && make

On my machine (ubuntu VM) I cannot reproduce on 0df6b9168e9dc09d3e0ae6050d67b7c31c603782 (Feb 22) nor a4c3ac2799281b56c1c38d0fdc584e22ac1b3d2f (latest). Steps I took:

```
nvim -u NONE -c 'set directory^=~/.nvim/swp//' -c 'e foo.txt'
<edit some text>
:preserve
kill -9 ...
nvim -u NONE -c 'set directory^=~/.nvim/swp//' -c 'e foo.txt'
(R)ecover
```

**Comment by justinmk** [comment]
@frangio  Correction, I can reproduce. I forgot to create `~/.nvim/swp/` so it was just using the current directory.

**Comment by justinmk** [comment]
Debugging this now.

**Comment by justinmk** [comment]
@frangio Thanks for the precise repro steps, I have a fix here: https://github.com/neovim/neovim/pull/2117

**Comment by frangio** [comment]
Excellent! Great work, Justin. :)

**Comment by dequis** [comment]
Uhhh, sorry, don't have time to recompile and test this one atm, but is it actually related to the bug I originally reported?

**Comment by justinmk** [comment]
@dequis Based on your stacktrace, no :/ I had some [weird behavior](https://github.com/chrisbra/Recover.vim/issues/30#issuecomment-52694254) with `set noshelltemp` combined with recover.vim, using gvim on Windows.

But I just noticed this line in your trace:

    #7  0x000000000056e9dc in shell (cmd=0x236d520 "TERM=vt100 LC_ALL=C vim -u NONE -es -V  'protocols/account.c'", extra_args=extra_args@entry=0x0, input=input@entry=0x0, len=0, output=output@entry=0x7fff27c58348,
    nread=nread@entry=0x7fff27c58350, silent=true, forward_output=false) at /home/dx/test/asd/bezel/neovim-git/src/neovim-git/src/nvim/os/shell.c:242

Notice how it's calling vim :) Hmmm... here's the relevant block of code: https://github.com/chrisbra/Recover.vim/blob/master/autoload/recover.vim#L100-L156

If I run `TERM=vt100 LC_ALL=C vim -u NONE -es -V  foo` in the terminal, it enters interactive ex mode. And, `system()` and `:!` in neovim do not support interactive shells (but `:terminal` will, although for the use case of recover.vim this entire 56-line hack would be unnecessary). 

So, this is the same issue as https://github.com/neovim/neovim/issues/1496. I'm going to close this to keep duplication down, but it's good to have extra information, so thanks for the report.

**Comment by justinmk** [comment]
@dequis by the way, if you build the latest (not my PR, just the laster master), what happens if you run this:

    call system('TERM=vt100 LC_ALL=C vim -u NONE -es -V neovim/CMakeLists.txt')

For me, it hangs for a couple seconds, then I regain control with no output from the command.

If you feel like it, perhaps you could also try recover.vim again to see if the behavior persists.

**Comment by dequis** [comment]
Hah, neat, thanks for looking at it! And I guess I should thank @frangio for helping getting this bug triaged.

I'll test that in a while and report back!

**Comment by dequis** [comment]
Okay that was a longer while than expected. Tried that command. It hung for two minutes (while i pressed random keys), only returned after i pressed ctrl-c, with no output from the command either. Seems like it does get stuck forever instead of just a couple of seconds.

I tried recover.vim again (with both new and old swap files) and it hangs the same way, but there's no crash anymore after the ctrl-c. The recovery dialog is the same one as the nvim built-in one, so recover.vim is still not working (i guess my ctrl-c kills it).

Still, the ugliest part of this bug was fixed at some point, so that counts as a win in my book. Thanks!
